### DISCUSSION TITLE
Host-to-Host：received AUTHENTICATION_FAILED notify error

### DISCUSSION THREAD

**Discussion by rangeihub** [strongswan/strongswan#2034] [Q&A]
reference:IPsec tunnel mode with X.509 certificates
sun:
pki --gen --type ed25519 --outform pem > sunKey.pem
pki --gen --type rsa --size 3072 > sunKey.der
pki --req --type priv --in sunKey.pem \
          --dn "C=CH, O=strongswan, CN=sun.strongswan.org" \
          --san sun.strongswan.org --outform pem > sunReq.pem
		  
pki --issue --cacert strongswanCert.pem --cakey strongswanKey.pem \
            --type pkcs10 --in sunReq.pem --serial 01 --lifetime 1826 \
            --outform pem > sunCert.pem
config:
config setup

conn %default
        ikelifetime=60m
        keylife=20m
        rekeymargin=3m
        keyingtries=1
        keyexchange=ikev2

conn host-host
        left=192.168.10.4
        leftcert=/var/certs/sunCert.pem
        leftid=@sun.strongswan.org
        leftfirewall=yes
        right=192.168.10.5
        rightid=@moon.strongswan.org
        auto=add
: RSA sunKey.pem
./ipsec.d/certs/sunCert.pem
./swanctl/x509/sunCert.pem

moon:
config setup

conn %default
        ikelifetime=60m
        keylife=20m
        rekeymargin=3m
        keyingtries=1
        keyexchange=ikev2

conn host-host
        left=192.168.10.5
        leftcert=/var/certs/moonCert.pem
        leftid=@moon.strongswan.org
        leftfirewall=yes
        right=192.168.10.4
        rightid=@sun.strongswan.org
        auto=add

root@OpenWrt:/etc# find -name "moonCert.pem"
./ipsec.d/certs/moonCert.pem

root@OpenWrt:/tmp/certs# swanctl --list-certs
List of X.509 End Entity Certificates

  subject:  "C=CH, O=strongswan, CN=moon.strongswan.org"
  issuer:   "C=CH, O=strongSwan, CN=strongSwan Root CA"
  validity:  not before Dec 08 11:35:02 2023, ok
             not after  Dec 07 11:35:02 2028, ok (expires in 1824 days)
  serial:    01
  altNames:  moon.strongswan.org
  flags:
  authkeyId: 0d:bc:a3:fe:8d:f3:a0:fa:98:b5:37:6a:e0:a2:5c:4b:53:bf:90:05
  subjkeyId: da:33:97:a5:09:67:be:46:c4:bb:bc:4d:d5:4d:5c:a2:40:cd:31:d0
  pubkey:    ED25519 256 bits, has private key
  keyid:     c6:18:e6:0b:62:82:dd:f5:b5:d0:c6:db:42:fd:2f:a6:0b:ce:1a:89
  subjkey:   da:33:97:a5:09:67:be:46:c4:bb:bc:4d:d5:4d:5c:a2:40:cd:31:d0

sun:
root@OpenWrt:/etc# ipsec up host-host
initiating IKE_SA host-host[3] to 192.168.10.5
generating IKE_SA_INIT request 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) N(FRAG_SUP) N(HASH_ALG) N(REDIR_SUP) ]
sending packet: from 192.168.10.4[500] to 192.168.10.5[500] (900 bytes)
received packet: from 192.168.10.5[500] to 192.168.10.4[500] (325 bytes)
parsed IKE_SA_INIT response 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) CERTREQ N(FRAG_SUP) N(HASH_ALG) N(CHDLESS_SUP) N(MULT_AUTH) ]
selected proposal: IKE:AES_CBC_128/HMAC_SHA2_256_128/PRF_HMAC_SHA2_256/ECP_256
received cert request for "CN=strongSwan CA"
received 1 cert requests for an unknown ca
sending cert request for "CN=strongSwan CA"
authentication of 'sun.strongswan.org' (myself) with ED25519 successful
sending end entity cert "C=CH, O=strongswan, CN=sun.strongswan.org"
establishing CHILD_SA host-host{2}
generating IKE_AUTH request 1 [ IDi CERT N(INIT_CONTACT) CERTREQ IDr AUTH SA TSi TSr N(MOBIKE_SUP) N(ADD_4_ADDR) N(ADD_6_ADDR) N(MULT_AUTH) N(EAP_ONLY) N(MSG_ID_SYN_SUP) ]
sending packet: from 192.168.10.4[4500] to 192.168.10.5[4500] (832 bytes)
received packet: from 192.168.10.5[4500] to 192.168.10.4[4500] (80 bytes)
parsed IKE_AUTH response 1 [ N(AUTH_FAILED) ]
received AUTHENTICATION_FAILED notify error
establishing connection 'host-host' failed
root@OpenWrt:/etc# 
root@OpenWrt:/etc# 
root@OpenWrt:/etc# swanctl --list-certs

List of X.509 End Entity Certificates

  subject:  "C=CH, O=strongswan, CN=sun.strongswan.org"
  issuer:   "C=CH, O=strongSwan, CN=strongSwan Root CA"
  validity:  not before Dec 06 23:28:22 2023, ok
             not after  Dec 05 23:28:22 2028, ok (expires in 1824 days)
  serial:    01
  altNames:  sun.strongswan.org
  flags:     
  authkeyId: 00:9f:f5:40:aa:d4:02:ac:5a:cb:9b:f6:6b:b4:22:b2:c3:65:d2:41
  subjkeyId: 10:f0:34:9c:9a:66:70:e0:bf:2b:a2:d3:f3:b3:d5:3c:5b:ce:c1:fb
  pubkey:    ED25519 256 bits, has private key
  keyid:     2c:36:d2:59:a4:e8:02:a7:2e:5c:7c:c7:75:cb:0e:31:2f:bc:f9:3a
  subjkey:   10:f0:34:9c:9a:66:70:e0:bf:2b:a2:d3:f3:b3:d5:3c:5b:ce:c1:fb

Please help check the configuration：
received AUTHENTICATION_FAILED notify error

**Comment by tobiasbrunner** [comment]
Read the log at the other end (i.e. where the error notify is generated).

**Reply by rangeihub** [reply]
Dec  9 14:08:46 07[ENC] <host-host|1> ENCRYPTED payload found, stop parsing                       
Dec  9 14:08:46 07[ENC] <host-host|1> process payload of type ENCRYPTED                           
Dec  9 14:08:46 07[ENC] <host-host|1> found an encrypted payload                                  
Dec  9 14:08:46 07[ENC] <host-host|1> parsing NOTIFY payload, 8 bytes left                        
Dec  9 14:08:46 07[ENC] <host-host|1>   parsing rule 0 U_INT_8                                    
Dec  9 14:08:46 07[ENC] <host-host|1>   parsing rule 1 FLAG                                       
Dec  9 14:08:46 07[ENC] <host-host|1>   parsing rule 2 RESERVED_BIT                               
Dec  9 14:08:46 07[ENC] <host-host|1>   parsing rule 3 RESERVED_BIT                               
Dec  9 14:08:46 07[ENC] <host-host|1>   parsing rule 4 RESERVED_BIT                               
Dec  9 14:08:46 07[ENC] <host-host|1>   parsing rule 5 RESERVED_BIT                               
Dec  9 14:08:46 07[ENC] <host-host|1>   parsing rule 6 RESERVED_BIT                               
Dec  9 14:08:46 07[ENC] <host-host|1>   parsing rule 7 RESERVED_BIT                               
Dec  9 14:08:46 07[ENC] <host-host|1>   parsing rule 8 RESERVED_BIT                               
Dec  9 14:08:46 07[ENC] <host-host|1>   parsing rule 9 PAYLOAD_LENGTH                             
Dec  9 14:08:46 07[ENC] <host-host|1>   parsing rule 10 U_INT_8                                   
Dec  9 14:08:46 07[ENC] <host-host|1>   parsing rule 11 SPI_SIZE                                  
Dec  9 14:08:46 07[ENC] <host-host|1>   parsing rule 12 U_INT_16                                  
Dec  9 14:08:46 07[ENC] <host-host|1>   parsing rule 13 SPI                                       
Dec  9 14:08:46 07[ENC] <host-host|1>   parsing rule 14 CHUNK_DATA                                
Dec  9 14:08:46 07[ENC] <host-host|1> parsing NOTIFY payload finished                             
Dec  9 14:08:46 07[ENC] <host-host|1> parsed content of encrypted payload                         
Dec  9 14:08:46 07[ENC] <host-host|1> insert decrypted payload of type NOTIFY at end of list      
Dec  9 14:08:46 07[ENC] <host-host|1> verifying message structure                                 
Dec  9 14:08:46 07[ENC] <host-host|1> found payload of type NOTIFY                                
Dec  9 14:08:46 07[ENC] <host-host|1> parsed IKE_AUTH response 1 [ N(AUTH_FAILED) ]               
Dec  9 14:08:46 07[IKE] <host-host|1> received AUTHENTICATION_FAILED notify error                 
Dec  9 14:08:46 07[CHD] <host-host|1> CHILD_SA host-host{1} state change: CREATED => DESTROYING   
Dec  9 14:08:46 07[KNL] <host-host|1> deleting SAD entry with SPI c982c1b4                        
Dec  9 14:08:46 07[KNL] <host-host|1> deleted SAD entry with SPI c982c1b4                         
Dec  9 14:08:46 07[MGR] <host-host|1> checkin and destroy IKE_SA host-host[1]                     
Dec  9 14:08:46 07[IKE] <host-host|1> IKE_SA host-host[1] state change: CONNECTING => DESTROYING  
Dec  9 14:08:46 07[MGR] checkin and destroy of IKE_SA successful

**Comment by rangeihub** [comment]
[charon.log](https://github.com/strongswan/strongswan/files/13658489/charon.log)
Are certificates generated on their respective hosts? not need to use the same key.pem?

**Reply by tobiasbrunner** [reply]
```
Dec  9 14:18:46 16[CFG] <host-host|2>   using certificate "C=CH, O=strongswan, CN=sun.strongswan.org"
Dec  9 14:18:46 16[CFG] <host-host|2>   certificate "C=CH, O=strongswan, CN=sun.strongswan.org" key: 256 bit ED25519
Dec  9 14:18:46 16[CFG] <host-host|2> no issuer certificate found for "C=CH, O=strongswan, CN=sun.strongswan.org"
Dec  9 14:18:46 16[CFG] <host-host|2>   issuer is "C=CH, O=strongSwan, CN=strongSwan Root CA"
Dec  9 14:18:46 16[IKE] <host-host|2> no trusted ED25519 public key found for 'sun.strongswan.org'
```
As you can see, the issuer certificate is missing.

>  Are certificates generated on their respective hosts?

You can generate a key on the respective host and then create a certificate request so the CA can issue a certificate. But you need to share the CA certificate between the hosts so they can trust all the issued certificates.  Alternatively, you could use self-signed certificates, but you then have to install these on all the hosts that communicate with each other. With the CA certificate you only need to share that (each host will then only need the CA certificate and their own key and certificate).

**Comment by rangeihub** [comment]
Key generation：
openssl genrsa -out ca-key.pem
openssl req -new -key ca-key.pem -out ca-req.pem
openssl x509 -req -in ca-req.pem -signkey ca-key.pem -out ca-cert.pem

openssl genrsa -out server-key.pem
openssl req -new -key server-key.pem -out server-req.pem
openssl x509 -req -in server-req.pem -CA ca-cert.pem -CAkey ca-key.pem -out server-cert.pem

openssl genrsa -out  client-key.pem
openssl req -new -key client-key.pem -out client-req.pem
openssl x509 -req -in client-req.pem -CA ca-cert.pem -CAkey ca-key.pem -out client-cert.pem

server:
config setup

conn %default
        ikelifetime=60m
        keylife=60m
        rekeymargin=3m
        keyingtries=1
        keyexchange=ikev1
        auto=start
        closeaction=hold
        dpdaction=hold

conn "ipsec"
        left=192.168.10.5
        right=192.168.10.4
        authby=rsasig
        leftcert=/var/openssl/server-cert.pem
        leftsendcert=always
        ikelifetime=3600s
        ike=aes128-sha1-modp1024
        lifetime=3600s
        esp=aes128-sha1
        leftfirewall=yes
        leftca=/var/openssl/ca-cert.pem

# /etc/ipsec.secrets - strongSwan IPsec secrets file
: RSA server-key.pem

[charon.log](https://github.com/strongswan/strongswan/files/13669420/charon.log)

May I ask if this configuration is correct? Unable to connect

**Comment by rangeihub** [comment]
I can establish a connection successfully when I provide the peer certificate. How should I configure it so that the peer certificate is not required?
